Sensors/SMCBatteryManager/SMCBatteryManager.cpp

```diff
@@ -85,10 +87,12 @@ bool SMCBatteryManager::start(IOService *provider) {
 		VirtualSMCAPI::addKey(KeyB0RM(i), vsmcPlugin.data, VirtualSMCAPI::valueWithUint16(2000, new B0RM(i), SMC_KEY_ATTRIBUTE_PRIVATE_WRITE|SMC_KEY_ATTRIBUTE_WRITE|SMC_KEY_ATTRIBUTE_READ));
 		VirtualSMCAPI::addKey(KeyB0St(i), vsmcPlugin.data, VirtualSMCAPI::valueWithData(nullptr, 2, SmcKeyTypeHex, new B0St(i), SMC_KEY_ATTRIBUTE_PRIVATE_WRITE|SMC_KEY_ATTRIBUTE_WRITE|SMC_KEY_ATTRIBUTE_READ));
 		VirtualSMCAPI::addKey(KeyB0TF(i), vsmcPlugin.data, VirtualSMCAPI::valueWithUint16(0, new B0TF(i)));
-		VirtualSMCAPI::addKey(KeyTB0T(i+1), vsmcPlugin.data, VirtualSMCAPI::valueWithSp(0, SmcKeyTypeSp78, new TB0T(i)));
+		if (BatteryManager::getShared()->state.btInfo[i].state.temperature != -20) {
```

Comment:
Why is -20°C an invalid temperature? If TjMAX is e.g. 99°C, then temperatures can be monitored from -28°C to +99°C. Also floating point comparisons are not reliable, although this particular case is OK (small integers are precisely represented as `double`, and the value `-20` is implicitly converted to `double` and becomes exactly the same as the initial value).

Reply:
I picked it from [Smart Battery Data Specification](http://sbs-forum.org/specs/sbdat110.pdf) `5.1.9 Temperature()`, typically the range is -20 to +75 Celsius. Can you suggest some values or approaches?

Reply:
Typical values are not allowed range. The same paragraph of the spec states that the range is 0°K to 6553.5°K. Either compare `temperatureKelvin` to 0 and check `BSSTemperatureSMBusOnly` again here, or add a flag.

Reply:
My bad. I thought `state` is unaccessible in `getBatteryInfo`, but it turned out it's `updateRealTimeStatus` that can only access `state`. Almost forgot `state` reside under `BatteryInfo` and tried to restrict it in the same level.

Reply:
> Why is -20°C an invalid temperature? If TjMAX is e.g. 99°C, then temperatures can be monitored from -28°C to +99°C. Also floating point comparisons are not reliable, although this particular case is OK (small integers are precisely represented as `double`, and the value `-20` is implicitly converted to `double` and becomes exactly the same as the initial value).

after upgrading to kext release, i don't read the temperatures anymore, I did in my fork the revert and I compiled kext, and now read back the temperatures, what breaks the reading in these changes?

Reply:
The change here is outdated. Please use bugtracker for issues other than DM and comments under commits. And you haven't attached troubleshooting files, specifically your ACPI implementation and the log from `SMCBatteryManager` and `SMCProcessor`.
A suspect for missing `TB0T` could be the non-zero check for temperature, resulting from slow firmware initialization. But I still have no idea why if will affect `SMCProcessor`.